src/audio/module_adapter/library/userspace_proxy.c

```diff
@@ -0,0 +1,462 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+// Author: Jaroslaw Stelter <jaroslaw.stelter@intel.com>
+// Author: Adrian Warecki <adrian.warecki@intel.com>
+#include <sof/common.h>
+#include <rtos/alloc.h>
+#include <rtos/cache.h>
+#include <sof/lib/memory.h>
+#include <rtos/string.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <sof/lib_manager.h>
+#include <sof/audio/component.h>
+#include <rtos/userspace_helper.h>
+#include <utilities/array.h>
+#include <zephyr/sys/sem.h>
+#include <sof/audio/module_adapter/module/generic.h>
+
+#include <sof/audio/module_adapter/library/userspace_proxy.h>
+#include <rimage/sof/user/manifest.h>
+
+LOG_MODULE_REGISTER(userspace_proxy_tr, CONFIG_SOF_LOG_LEVEL);
+
+/* 6f6b6f4b-6f73-7466-20e1e62b9779f003 */
+SOF_DEFINE_REG_UUID(userspace_proxy);
+
+DECLARE_TR_CTX(userspace_proxy_tr, SOF_UUID(userspace_proxy_uuid), LOG_LEVEL_INFO);
+
+static const struct module_interface userspace_proxy_interface;
+
+static int userspace_proxy_memory_init(struct userspace_context *user,
+				       const struct comp_driver *drv)
+{
+	/* Add module private heap to memory partitions */
+	struct k_mem_partition heap_part = { .attr = K_MEM_PARTITION_P_RW_U_RW };
+	k_mem_region_align(&heap_part.start, &heap_part.size,
+			   POINTER_TO_UINT(drv->user_heap->init_mem),
+			   DRV_HEAP_SIZE, CONFIG_MM_DRV_PAGE_SIZE);
+
+#if !defined(CONFIG_XTENSA_MMU_DOUBLE_MAP) && defined(CONFIG_SOF_ZEPHYR_HEAP_CACHED)
+#define HEAP_PART_CACHED
+	/* Add cached module private heap to memory partitions */
+	struct k_mem_partition heap_part_cached = { .attr = K_MEM_PARTITION_P_RW_U_RW };
+	k_mem_region_align(&heap_part_cached.start, &heap_part_cached.size,
+			   POINTER_TO_UINT(sys_cache_cached_ptr_get(drv->user_heap->init_mem)),
+			   DRV_HEAP_SIZE, CONFIG_MM_DRV_PAGE_SIZE);
+#endif
+
+	struct k_mem_partition *parts_ptr[] = {
+#ifdef HEAP_PART_CACHED
+		&heap_part_cached,
+#endif
+		&heap_part
+	};
+
+	return k_mem_domain_init(user->comp_dom, ARRAY_SIZE(parts_ptr), parts_ptr);
+}
+
+static int userspace_proxy_add_sections(struct userspace_context *user, uint32_t instance_id,
+					 const struct sof_man_module *const mod)
+{
+	uint32_t mem_partition;
+	void* va_base;
+	size_t size;
+	int idx, ret;
+
+	for (idx = 0; idx < ARRAY_SIZE(mod->segment); ++idx) {
+		if (!mod->segment[idx].flags.r.load)
+			continue;
+
+		mem_partition = K_MEM_PARTITION_P_RO_U_RO;
+
+		if (mod->segment[idx].flags.r.code)
+			mem_partition = K_MEM_PARTITION_P_RX_U_RX;
+		else if (!mod->segment[idx].flags.r.readonly)
+			mem_partition = K_MEM_PARTITION_P_RW_U_RW;
+
+		va_base = UINT_TO_POINTER(mod->segment[idx].v_base_addr);
+		size = mod->segment[idx].flags.r.length * CONFIG_MM_DRV_PAGE_SIZE;
+
```

Comment:
If `length` is corrupted we can have integer overflow. (To check - 16 bits for length)